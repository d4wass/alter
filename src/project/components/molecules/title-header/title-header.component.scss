@use '../../../../base';

.wrapper {
  padding: 20px 30px;
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 50vh;
  position: relative;
  background-size: cover;
}

.main {
  background-image: url('../../../../assets/header.png');
}

.faq {
  background-image: url('../../../../assets/learn-more-header.jpg');
  background-position-y: 30%;
  .title-container {
    bottom: 30%;
    padding: 0;
  }
}

.host {
  background-image: url('../../../../assets/host.jpg');
  background-position-y: 55%;
  .title-container {
    bottom: 30%;
    padding: 0;
  }
}

.user {
  background-image: url('../../../../assets/user.jpg');
  background-position-y: 45%;
  .title-container {
    bottom: 30%;
    padding: 0;
  }
}

.title-container {
  background-color: #fff;
  padding: 0 50px 50px;
  border-radius: base.$main-border-rad;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  position: absolute;
  bottom: -8%;
  width: 70%;
}

.title {
  font-size: 5em;
  font-weight: 900;
  text-align: center;
}