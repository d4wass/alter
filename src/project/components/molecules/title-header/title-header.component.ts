import { Component, Input, OnInit } from '@angular/core';
import { Router } from '@angular/router';

@Component({
    selector: 'app-title-header',
    template: `
    <div class="wrapper" [ngClass]="bgHeader">
      <div class="title-container">
        <h1 class="title">{{ title }}</h1>
        <ng-container *ngIf="isRouter">
          <app-main-search-form></app-main-search-form>
        </ng-container>
      </div>
    </div>
  `,
    styleUrls: ['./title-header.component.scss'],
    standalone: false
})
export class TitleHeaderComponent implements OnInit {
  @Input() title!: string;
  isRouter!: boolean;
  bgHeader!: string;

  constructor(private _router: Router) {}

  ngOnInit(): void {
    this.isRouter = this._router.url === '/';
    this.setBgHeader();
  }

  private setBgHeader(): void {
    switch (this._router.url) {
      case '/':
        this.bgHeader = 'main';
        break;
      case '/faq':
        this.bgHeader = 'faq';
        break;
      case '/host':
        this.bgHeader = 'host';
        break;
      case '/new-user':
        this.bgHeader = 'user';
        break;
      default:
        break;
    }
  }
}
