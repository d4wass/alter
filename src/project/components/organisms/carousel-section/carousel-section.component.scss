:host{

}

.carousel-section{
    padding: 50px 0;
    h1{
        margin-bottom: 50px;
    }
}

.carousel-content {
    display: flex;
    justify-content: center;
    align-items: center;
    img {
        height: 50px;
    }

    .car-card-wrapper {
        width: 100%;
        padding: 0 30px;
        display: flex;
        justify-content: space-between;
    }
}