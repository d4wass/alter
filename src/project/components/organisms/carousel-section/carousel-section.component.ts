import { animate, AnimationBuilder, AnimationPlayer, style } from '@angular/animations';
import {
  AfterViewInit,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ElementRef,
  OnInit,
  QueryList,
  Renderer2,
  ViewChild,
  ViewChildren
} from '@angular/core';

const carCards = [
  { title: 'BMW', img: 'assets/car.png' },
  { title: 'Toyota', img: 'assets/car.png' },
  { title: 'Mazda', img: 'assets/car.png' },
  { title: 'Jeep', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' }
];

@Component({
  selector: 'app-carousel-section',
  template: `
    <div class="carousel-section">
      <h1>Browse by make</h1>
      <div class="carousel-content">
        <div (click)="handleClick('-')" class="nav-btn">
          <img src="assets/angle-left.svg" />
        </div>
        <div class="car-card-wrapper" #carousel>
          <app-car-card
            *ngFor="let item of carCards"
            #carouselCards
            [carCard]="item"
          ></app-car-card>
        </div>
        <div (click)="handleClick('')" class="nav-btn">
          <img src="assets/angle-right.svg" />
        </div>
      </div>
    </div>
  `,
  styleUrls: ['./carousel-section.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CarouselSectionComponent implements AfterViewInit {
  carCards = carCards;
  time = '250ms ease-in';
  private player!: AnimationPlayer;
  private carouselItems!: HTMLElement[];
  private carouselWidth!: number;
  private counter = 0;
  private offset: number = 0;

  constructor(
    private animationBuilder: AnimationBuilder,
    private cdr: ChangeDetectorRef,
    private renderer: Renderer2
  ) {}
  //todo: whats the difference between viewChild and ViewChildren
  @ViewChild('carousel') private carousel!: ElementRef;
  @ViewChildren('carouselCards') private carouselCards!: QueryList<ElementRef>;

  ngAfterViewInit(): void {
    this.carouselItems = [...this.carousel.nativeElement.children];
    setTimeout(() => (this.carouselWidth = this.carousel.nativeElement.clientWidth));
  }

  //add proper types for all
  handleClick(direction: string): void {
    const itemWidth = this.carouselItems[0].clientWidth;
    const maxOffset = this.carouselItems.length * itemWidth - this.carouselWidth;
    this.offset = this.checkCurrentOffset(direction, itemWidth, maxOffset);
    const carouselAnimation = this.buildAnimation();

    this.carouselItems.forEach((item) => {
      this.player = carouselAnimation.create(item);
      this.player.onDone(() => {});
      this.player.play();
    });
  }

  private buildAnimation() {
    return this.animationBuilder.build([
      animate(this.time, style({ transform: `translateX(${this.offset}px)` }))
    ]);
  }

  private counterDirectionControl(direction: string, maxCounterValue: number): void {
    let counter = this.counter;
    if (direction === '' && this.counter !== 0) {
      --counter;
      this.counter = counter;
    }

    if (direction === '-' && this.counter === maxCounterValue) {
      this.counter = counter;
    } else if (direction === '-') {
      ++counter;
      this.counter = counter;
    }
  }

  private checkCurrentOffset(direction: string, itemWidth: number, maxOffset: number): number {
    let newOffset = this.offset;
    const num = this.carouselWidth / this.carouselItems[0].clientWidth;
    const roundedNum = this.roundNumber(num); // how many entire cards I can display for user
    const numberOfCardsVisibleOnCarousel = this.carouselItems.length - Math.floor(roundedNum) - 1;
    const numberOfNonVisibleCardsOnCarousel =
      this.carouselItems.length - numberOfCardsVisibleOnCarousel;
    const movePxOfLastCard = this.checkIfPartOfItemIsVisible(
      numberOfCardsVisibleOnCarousel,
      itemWidth
    );
    this.counterDirectionControl(direction, numberOfNonVisibleCardsOnCarousel);
    console.log(
      'counter',
      this.counter,
      'number of visible cards',
      numberOfCardsVisibleOnCarousel,
      'how many px move last card',
      movePxOfLastCard,
      'non visible cards on carousel',
      numberOfNonVisibleCardsOnCarousel
    );

    //! rework this conditions
    if (direction === '' && newOffset === 0) {
      console.log('1 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = 0);
    }
    if (this.counter === numberOfNonVisibleCardsOnCarousel) {
      console.log('3 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = newOffset - movePxOfLastCard);
    }
    if (direction === '-' && newOffset <= 0) {
      console.log('2 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = newOffset - itemWidth);
    }
    if (direction === '-' && newOffset <= -maxOffset) {
      console.log('4 offset set offset to max valuxe');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = -maxOffset);
    }
    if (direction === '' && newOffset > 0) {
      console.log('5 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = 0);
    }

    return direction ? newOffset - itemWidth : newOffset + itemWidth;
  }

  private roundNumber(num: number): number {
    return Math.round((num + Number.EPSILON) * 100) / 100;
  }

  private checkIfPartOfItemIsVisible(numberOfVisibleCards: number, itemWidth: number): number {
    // after implement should return number
    console.log(this.carouselWidth);
    const visibleItemPartWidth = this.carouselWidth - numberOfVisibleCards * itemWidth;
    const movePx = itemWidth - visibleItemPartWidth;
    return movePx;
  }
}
