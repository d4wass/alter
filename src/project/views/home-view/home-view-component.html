<main>
  <app-title-header [title]="'Find your drive'"></app-title-header>
  <div class="flex-wrapper">
    <app-carousel-section
      [carouselItems]="cars$ | async"
      [carouselTitle]="'Browse by make'"
      [carouselType]="CarouselEnum.CarCarousel"
    ></app-carousel-section>
    <app-carousel-section
      [carouselItems]="reviews"
      [carouselTitle]="'Alter Hosts'"
      [carouselType]="CarouselEnum.ReviewCarousel"
    ></app-carousel-section>
    <app-faq-section></app-faq-section>
  </div>
</main>
