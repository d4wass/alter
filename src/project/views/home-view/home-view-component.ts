import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { map, Observable, take, tap } from 'rxjs';
import { AppSettingFacade } from '../../../+state/facade/app-state/app-settings.facade';
import { CarouselEnum } from '../../components/organisms/carousel-section/carousel.utils';

const carsData = [
  { title: '1', img: 'assets/car.png' },
  { title: '2', img: 'assets/car.png' },
  { title: '3', img: 'assets/car.png' },
  { title: '4', img: 'assets/car.png' },
  { title: '5', img: 'assets/car.png' },
  { title: '6', img: 'assets/car.png' },
  { title: '7', img: 'assets/car.png' },
  { title: '8', img: 'assets/car.png' },
  { title: '9', img: 'assets/car.png' },
  { title: '10', img: 'assets/car.png' },
  { title: '11', img: 'assets/car.png' }
];

const reviewsData = [
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 10, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 3
  },
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 20, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 4
  },
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 30, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 5
  },
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 40, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 2
  },
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 30, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 5
  },
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 40, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 2
  },
  {
    img: 'assets/reviewUser.png',
    name: 'Name',
    surname: 'Surname',
    status: 'All-star Host',
    stats: { trips: 40, joined: 'Jan 2016' },
    review:
      'Ran well. Very convenient check in and check out. I was using it to move some small pelican cases, so loading into th...',
    rate: 2
  }
];

@Component({
  selector: 'app-home-view-component',
  templateUrl: './home-view-component.html',
  styleUrls: ['./home-view-component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class HomeViewComponent implements OnInit {
  constructor(private appSettingsFacade: AppSettingFacade) {}
  cars$!: Observable<{ title: string; img: string }[]>;

  ngOnInit(): void {
    this.cars$ = this.appSettingsFacade.vehicleBrands$.pipe(
      map((brands) => brands.map((brand) => ({ title: brand, img: 'assets/car.png' }))),
      tap((brands) => console.log(brands))
    );
  }

  reviews = reviewsData;
  // cars = carsData;
  CarouselEnum = CarouselEnum;
}
