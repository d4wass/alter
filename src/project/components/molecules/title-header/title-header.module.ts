import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TitleHeaderComponent } from './title-header.component';
import { MainSearchFormModule } from '../main-search-form/main-search-form.module';

@NgModule({
  declarations: [TitleHeaderComponent],
  imports: [CommonModule, MainSearchFormModule],
  exports: [TitleHeaderComponent]
})
export class TitleHeaderModule {}
