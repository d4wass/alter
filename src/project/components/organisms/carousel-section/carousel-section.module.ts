import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CarouselSectionComponent } from './carousel-section.component';
import { CarCardComponent } from '../../molecules/car-card/car-card.component';

@NgModule({
  declarations: [CarouselSectionComponent, CarCardComponent],
  imports: [CommonModule],
  exports: [CarouselSectionComponent]
})
export class CarouselSectionModule {}
