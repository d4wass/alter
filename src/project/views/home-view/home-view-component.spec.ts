import { createComponentFactory, Spectator } from '@ngneat/spectator';
import { MockModule } from 'ng-mocks';
import { of } from 'rxjs';
import {
  AppSettingsFacadeMock,
  AppSettingsFacadeMockProvider
} from '../../../+state/facade/app-state/app-state.facade.mock';
import { TitleHeaderModule } from '../../../project/components/molecules/title-header/title-header.module';
import { CarouselSectionModule } from '../../../project/components/organisms/carousel-section/carousel-section.module';
import { FaqSectionModule } from '../../../project/components/organisms/faq-section/faq-section.module';
import { HomeViewComponent } from './home-view-component';

describe('HomeViewComponentComponent', () => {
  let component: HomeViewComponent;
  let spectator: Spectator<HomeViewComponent>;
  let appFacade: AppSettingsFacadeMock;

  const createComponent = createComponentFactory({
    component: HomeViewComponent,
    shallow: true,
    imports: [
      MockModule(TitleHeaderModule),
      MockModule(FaqSectionModule),
      MockModule(CarouselSectionModule)
    ],
    providers: [AppSettingsFacadeMockProvider]
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
    appFacade = spectator.inject(AppSettingsFacadeMock);
  });

  it('should match to snapshot and be defined', () => {
    expect(spectator.fixture).toMatchSnapshot();
    expect(component).toBeTruthy();
  });

  describe('ngOnInit', () => {
    it('should set cars$ value on init', () => {
      const vehicleBrands = ['Car 1', 'Car 2'];
      const expected = vehicleBrands.map((brand) => ({ title: brand, img: 'assets/car.png' }));
      appFacade.vehicleBrands$.next(vehicleBrands);

      component.ngOnInit();

      component.cars$.subscribe((cars) => {
        expect(cars).toEqual(expected);
      });
    });
  });
});
