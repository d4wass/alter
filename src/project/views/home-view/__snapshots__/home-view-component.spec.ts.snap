// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeViewComponentComponent should match to snapshot and be defined 1`] = `
<app-home-view-component
  CarouselEnum={[Function Object]}
  appSettingsFacade={[Function AppSettingsFacadeMock]}
  cars$={[Function AnonymousSubject]}
  reviews={[Function Array]}
>
  <main>
    <app-title-header />
    <div
      class="flex-wrapper"
    >
      <app-carousel-section />
      <app-carousel-section />
      <app-faq-section />
    </div>
  </main>
</app-home-view-component>
`;
